### PR TITLE
userDto added;

### PR DESCRIPTION
**Pull request by kmety**
registration with fields validation added;
Role entity, dao, service added;
userDetailsService added.

### CODE REVIEW

File: src/main/java/mate/academy/spring/controller/RentController.java [hunk 1]
```diff
@@ -53,19 +60,32 @@ public String returnBook(@RequestParam("book_id") Long bookId, Model model) {
...
             return "errorPage";
         }
         List<Book> books = rentService.getBooksRentedByUser(optionalUser.get());
         model.addAttribute("books", books);
-        return "books";
+        return "rentedBooks";
+    }
+
+    private Optional<User> getCurrentUser(Principal principal) {
+        if (principal == null) {
```
Comment (Sofasmile) [inline src/main/java/mate/academy/spring/controller/RentController.java]: do you think that principal can be null?

File: src/main/java/mate/academy/spring/service/impl/UserServiceImp.java [hunk 2]
```diff
@@ -15,14 +18,52 @@
...
+                    "There is an account with that email adress: "
+                            + userDto.getEmail());
+        }
+        if (usernameExist(userDto.getUserName())) {
+            throw new IllegalArgumentException(
+                    "There is an account with that username: "
+                            + userDto.getUserName());
+        }
+        User user = new User();
+        user.setName(userDto.getFirstName());
```
Comment (Sofasmile) [inline src/main/java/mate/academy/spring/service/impl/UserServiceImp.java]: better create separate method for mapping dto to entity in DtoUtil class

File: src/main/java/mate/academy/spring/config/SecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    @Autowired
+    public void configureGlobal(AuthenticationManagerBuilder auth) throws Exception {
+        auth
+                .userDetailsService(userDetailsService()).passwordEncoder(passwordEncoder());
+    }
+
+    protected void configure(HttpSecurity http) throws Exception {
+        http
+                .authorizeRequests()
+                .antMatchers("/book/add").hasRole("ADMIN")
```
[the remark this answers is not in the document]
Comment (kmety) [reply src/main/java/mate/academy/spring/config/SecurityConfig.java]: done

File: src/main/java/mate/academy/spring/dao/impl/UserDaoImp.java [hunk 4]
```diff
@@ -17,10 +17,39 @@
 
     @Override
     public User add(User user) {
+        System.out.println(user.getRoles());
         sessionFactory.getCurrentSession().save(user);
         return user;
     }
 
+    @Override
+    public Optional<User> getByUsername(String username) {
```
[the remark this answers is not in the document]
Comment (kmety) [reply src/main/java/mate/academy/spring/dao/impl/UserDaoImp.java]: done

File: src/main/java/mate/academy/spring/controller/RentController.java (hunk 1 above)
Comment (kmety) [reply src/main/java/mate/academy/spring/controller/RentController.java]: Действительно, не может. Даже если незарегистрированный пользователь явно напишет запрос: "http://localhost:8080/rent/rentBook?book_id=1", спринг секьюрити перенаправит его на страницу логина. Проверка на null здесь не нужна

File: src/main/java/mate/academy/spring/service/impl/UserServiceImp.java (hunk 2 above)
Comment (kmety) [reply src/main/java/mate/academy/spring/service/impl/UserServiceImp.java]: done

File: src/main/java/mate/academy/spring/dto/DtoUtil.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+    public User parse(UserRegistrationDto userDto) {
+        User user = new User();
+        user.setName(userDto.getFirstName());
+        user.setSurname(userDto.getLastName());
+        user.setEmail(userDto.getEmail());
+        user.setUsername(userDto.getUserName());
+        user.setPassword(passwordEncoder.encode(userDto.getPassword()));
+        Role role;
+        if ("ROLE_ADMIN".equals(userDto.getRole())) {
```
Comment (Sofasmile) [inline src/main/java/mate/academy/spring/dto/DtoUtil.java]: do we need this checking?
better write:
Role role = roleService.getByName(userDto.getRole());

File: src/main/java/mate/academy/spring/service/impl/RoleServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+public class RoleServiceImpl implements RoleService {
+    @Autowired
+    private RoleDao roleDao;
+
+    @Transactional
+    @Override
+    public void addRole(String roleName) {
```
[the remark this answers is not in the document]
Comment (Sofasmile) [reply src/main/java/mate/academy/spring/service/impl/RoleServiceImpl.java]: `add` role means `create` new role. If we want to add new role like MODERATOR, we should firstly add this role in program code and then  add it in db. sounds not good
i think we can remove list of roles in line 15)

File: src/main/java/mate/academy/spring/service/impl/RoleServiceImpl.java [hunk 7]
```diff
@@ -9,39 +12,28 @@
...
-        switch (roleName.toLowerCase()) {
-            case ("user"):
-                roleName = "ROLE_USER";
-                break;
-            case ("admin"):
-                roleName = "ROLE_ADMIN";
-                break;
-            default:
-                throw new IllegalArgumentException("Wrong role name");
+        if (ROLE_NAMES.contains(roleName)) {
```
Comment (Sofasmile) [inline src/main/java/mate/academy/spring/service/impl/RoleServiceImpl.java]: the same here

Comment (kmety) [reply src/main/java/mate/academy/spring/service/impl/RoleServiceImpl.java]: agree. fixed

**Pull request merged**